### ISSUE TITLE
Does rxdart Observable want to follow the standard ReaxtiveX onError contract?

### ISSUE THREAD

**Issue by lestathc** [ReactiveX/rxdart#68]
Hi, 

I am very interesting in this lib, this is a big improvement for Dart.

But I have a concern:

1. as described in the [contract of observable](http://reactivex.io/documentation/contract.html), the onError will terminate the subscription. 

2. but as described in the [Stream.listen](https://api.dartlang.org/stable/1.22.0/dart-async/Stream/listen.html), the `cancelOnError` is false (by default), subscription will still exist.

How does this lib want to deal this case?

Thanks

**Comment by brianegan** [comment]
This is a good Q, and something we should document. Good catch, and certainly a tension between two worlds.

What do you think about this? Would love to hear your opinion.

(For what it's worth, I'd prefer to stay more in line with Dart Streams rather than Observables, as I want to fit into that ecosystem as smoothly as possible).

Overall, would love to hear your opinion, and will definitely add docs around this! Thanks for writing in!

**Comment by lestathc** [comment]
I don't know the solution neither. This is not an implementation problem.

I agree with you that stay with Dart Stream (given it is in the Dart ecosystem) and document it. (but... I really really dislike the 'use with care' in document, like another [bug](https://github.com/ReactiveX/rxdart/issues/67) I posted to this lib, SyncStreamController could solve that but SyncStreamController is documented with ['it must be used carefully'](https://api.dartlang.org/be/144789/dart-async/SynchronousStreamController-class.html))

And if `cancelOnError` is false, what is the differences with the [`retry` operator](http://reactivex.io/documentation/operators/retry.html), user may be confused with it. (And also, I think the [retry in this lib](https://github.com/ReactiveX/rxdart/blob/master/lib/src/transformers/retry.dart) is not what the retry documented in the ReactiveX, this lib's retry did not **re-trigger** the onListen/onSubscribe side effect.)

There are also some other concepts are different between Stream and ReactiveX. Like:
* `first`, `last`... operators: in Stream it is a Future, but in ReactiveX it is a Observable;
* the ReactiveX Observable could be multi-subscribed, but non broadcasting Stream is not;
* not like ReactiveX, Stream exposed back pressure to the StreamSubscription (the `pause`, `resume` method), it needs more consideration when implementing operators.

I feel that there is a big gap between Stream and standard Observable, maybe more documents are needed.

**Comment by frankpepermans** [comment]
Retry indeed is off and needs a rework, would you mind creating a separate
issue for that one?

As for Stream vs Observable, Stream is a core Dart feature that is very
hard to move around. We could ditch Stream for example, but then you lose
any exchange between the two as well, would be a hard sell if you are used
to Dart only.

Let's keep this one open, I'm open to ideas :)

**Comment by lestathc** [comment]
I created a new bug for `retry`.

Back to this thread, I can say that extending Stream has many benefits, like in AngularDart, the events in widget are presented as a Stream, and Stream can be rendered to widget using AsyncPipe directly. It is a really positive example to let Observable extend the Stream. 

But personally, I also feel that the Stream is too heavy and many concepts are not very fit the Observable.

I am also looking for other's opinion.

**Comment by brianegan** [comment]
"Stream is too heavy and many concepts are not very fit the Observable."

I have a bit more relaxed opinion on this. 

Overall, I completely agree there are some subtle but important differences, but I actually don't think the situation is that bad. Of course, our Streams and Operators should be as close to the ReactiveX specs as possible, but overall, I'm actually not as worried about the inconsistencies you mention.

For example, RxJava1 did in fact return an Observable from `first` and `last`, but RxJava2 now returns a `Single` for those methods. `Single` is almost identical to `Future`.

Therefore, you could say the Dart Stream implementation was ahead of its time! 

Jokes aside, that's a big API change, and I think the amount of API change that other implementations are going through means that trying to mimic or keep up with others (which one would we keep up with?) would be unproductive. 

I'd rather we appreciate the subtle differences, document the important ones, and make the rest of RxDart as compatible with ReactiveX as we can within the context of Dart Stream compatibility.

Each environment has their own quirks. And that's ok.

**Comment by lestathc** [comment]
I think for RxJava2, `Single` is the optimized version for 1 object `Observable`, similar (`Flowable` is the optimized version for 10K+ objects `Observable`), they can be converted to each other fluently, e.g.:
```
single.concat => Observable
observable.single => Single
```
And I don't think `Single` is almost identical to `Future` since `Single` follows the [reactive-stream publisher spec](https://github.com/reactive-streams/reactive-streams-jvm#1-publisher-code), in another words `Single` is an Observable as well, but dart `Future` follows the [promise sepc](https://promisesaplus.com/) (I am not sure, but seems). There are many differences between these two specs.

But I agree with you that different implementation should do the trade off, Stream has its advantages, and you just need to decide which one should be kept and which one should follow the Observable. Documenting them is very important for users. 

Sorry for the nit picky, I do think reactive programming is very important for Dart and appreciate your contributions :)

**Comment by brianegan** [comment]
Hey, no worries, these details are super important. Sure, I was joking a bit about them being identical -- but the overall point was what was important -- other APIs change, and we should embrace them as best we can where it makes sense in the Dart ecosystem.

**Comment by brianegan** [comment]
One question I thought of though: We could very easily create `Observable` versions of `first`, `last`, etc. Do you think that would be a good idea @lestathc? Thanks again for your thoughts!

**Comment by lestathc** [comment]
Given `Observable` extends `Stream`, the return type of `first` and `last` should be subclass of `Future`. How do you solve this problem?

**Comment by frankpepermans** [comment]
First and last would be kept as-is. For convenience, something extra like firstAsStream and lastAsStream could be added.

Though today, it is just a one liner to wrap it using new Stream.fromFuture(stream.first)

**Comment by lestathc** [comment]
`firstAsStream` interface looks good to me!

**Comment by brianegan** [comment]
PR #80 opened to address the conversion from Future back to Observable. Let me know whatcha think @lestathc

**Comment by brianegan** [comment]
I think this issue has come to a close. To summarize the points and actions:

  - We've documented our decision to stick with `cancelOnError: false` by default in the Documentation for the Observable class
  - Added docs about Dart streams being single-subscription, ensured all are single-subscription
  - Ensured all our StreamTransformers can be re-used.
  - Single and Future have many implementation differences, but are conceptually similar. We've implemented an `AsObservableFuture` that is now returned from all `Future` returning methods that allows for easy transformations from Future back to Observable in a chain. e.g. `new Observable.fromIterable([1, 2, 3]).last.asObservable.listen(print); // prints 3`
  - Retry was broken (now fixed)
